pkg/storage/ingest/writer.go

```diff
@@ -189,38 +194,52 @@ func (w *Writer) WriteSync(ctx context.Context, partitionID int32, userID string
 // or an error occurred.
 func (w *Writer) produceSync(ctx context.Context, client *kgo.Client, records []*kgo.Record) kgo.ProduceResults {
 	var (
-		remaining = atomic.NewInt64(int64(len(records)))
-		done      = make(chan struct{})
-		resMx     sync.Mutex
-		res       kgo.ProduceResults
+		remaining        = atomic.NewInt64(int64(len(records)))
+		done             = make(chan struct{})
+		resMx            sync.Mutex
+		res              kgo.ProduceResults
+		maxBufferedBytes = int64(w.kafkaCfg.ProducerMaxBufferedBytes)
 	)
 
 	w.writeRequestsTotal.Add(float64(len(records)))
 
+	onProduceDone := func(r *kgo.Record, err error) {
+		if maxBufferedBytes > 0 {
+			w.writersBufferedBytes.Add(-int64(len(r.Value)))
+		}
+
+		resMx.Lock()
+		res = append(res, kgo.ProduceResult{Record: r, Err: err})
+		resMx.Unlock()
+
+		if err != nil {
+			w.writeFailuresTotal.WithLabelValues(produceErrReason(err)).Inc()
+		}
+
+		// In case of error we'll wait for all responses anyway before returning from produceSync().
+		// It allows us to keep code easier, given we don't expect this function to be frequently
+		// called with multiple records.
+		if remaining.Dec() == 0 {
+			close(done)
+		}
+	}
+
 	for _, record := range records {
-		// We use a new context to avoid that other TryProduce() may be cancelled when this call's context is
-		// canceled. It's important to note that cancelling the context passed to TryProduce() doesn't actually
+		// Fast fail if the Kafka client buffer is full.
```

Comment:
We can add comment that `Add` will be reversed in `onProduceDone`.

Reply:
Done in https://github.com/grafana/mimir/pull/8738/commits/c74391199b1270bc2f9bdb5ae55c1ca79c18d30e